### PR TITLE
[Merged by Bors] - chore: reenable eta, bump to nightly 2023-05-16

### PR DESCRIPTION
**Pull request by gebner**
Now that leanprover/lean4#2210 has been merged, this PR:
* removes all the `set_option synthInstance.etaExperiment true` commands (and some `etaExperiment%` term elaborators)
* removes many but not quite all `set_option maxHeartbeats` commands
* makes various other changes required to cope with leanprover/lean4#2210.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by gebner** [comment]
!bench

**Comment by gebner** [comment]
!bench

**Comment by gebner** [comment]
!bench

**Comment by semorrison** [comment]
!bench

File: lean-toolchain [hunk 1]
```diff
@@ -1 +1 @@
-leanprover/lean4:nightly-2023-05-06
+gebner/lean4:reenableeta230506
```
Comment (Parcly-Taxel) [inline lean-toolchain:1]:
```suggestion
leanprover/lean4:nightly-2023-05-16
```

Comment (semorrison) [reply lean-toolchain:1]: Patience. :-) That nightly won't exist for another 11 hours or so!

File: Mathlib/CategoryTheory/Limits/Shapes/Terminal.lean [hunk 2]
```diff
@@ -462,6 +462,11 @@ theorem hasInitial_of_hasTerminal_op [HasTerminal Cᵒᵖ] : HasInitial C :=
   (initialUnopOfTerminal terminalIsTerminal).hasInitial
 #align category_theory.limits.has_initial_of_has_terminal_op CategoryTheory.Limits.hasInitial_of_hasTerminal_op
 
+-- Porting note:
+-- Workaround for issue discussed at https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/Failure.20of.20TC.20search.20in.20.60simp.60.20with.20.60etaExperiment.60.2E
+-- when etaExperiment is on.
+attribute [local aesop safe (rule_sets [CategoryTheory])] Subsingleton.elim
+
```
Comment (semorrison) [inline Mathlib/CategoryTheory/Limits/Shapes/Terminal.lean]: This has already been added globally to the rule set, so I've removed this change.

File: Mathlib/Data/Fin/Basic.lean [hunk 3]
```diff
@@ -986,6 +986,8 @@ theorem lt_add_one_iff {n : ℕ} {k : Fin (n + 1)} : k < k + 1 ↔ k < last n :=
   simp
 #align fin.lt_add_one_iff Fin.lt_add_one_iff
 
+attribute [nolint simpNF] lt_add_one_iff add_one_lt_iff -- HACK: CovariantClass lemma times out sigh
+
```
Comment (semorrison) [inline Mathlib/Data/Fin/Basic.lean:990]: I guess we better test this again, but I'm happy if this PR goes through before we get to that.

File: Mathlib/AlgebraicTopology/SimplexCategory.lean [hunk 4]
```diff
@@ -583,8 +583,10 @@ theorem iso_eq_iso_refl {x : SimplexCategory} (e : x ≅ x) : e = Iso.refl x :=
   have eq₁ := Finset.orderEmbOfFin_unique' h fun i => Finset.mem_univ ((orderIsoOfIso e) i)
   have eq₂ :=
     Finset.orderEmbOfFin_unique' h fun i => Finset.mem_univ ((orderIsoOfIso (Iso.refl x)) i)
-  ext1; ext1
-  exact congr_arg (fun φ => OrderEmbedding.toOrderHom φ) (eq₁.trans eq₂.symm)
+  ext x
+  replace eq₁ := congr_arg (· x) eq₁
+  replace eq₂ := congr_arg (· x) eq₂.symm
+  simp_all
```
Comment (Ruben-VandeVelde) [inline Mathlib/AlgebraicTopology/SimplexCategory.lean:589]: Should this get a porting note?

File: Mathlib/Data/Dfinsupp/Order.lean [hunk 5]
```diff
@@ -63,7 +63,8 @@ def orderEmbeddingToFun : (Π₀ i, α i) ↪o ∀ i, α i where
 #align dfinsupp.order_embedding_to_fun Dfinsupp.orderEmbeddingToFun
 
 @[simp]
-theorem orderEmbeddingToFun_apply {f : Π₀ i, α i} {i : ι} : orderEmbeddingToFun f i = f i :=
+theorem orderEmbeddingToFun_apply {f : Π₀ i, α i} {i : ι} :
+    (@orderEmbeddingToFun ι α _ _ f) i = f i :=
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Dfinsupp/Order.lean:67]: Porting note?

File: Mathlib/Data/Polynomial/FieldDivision.lean [hunk 6]
```diff
@@ -366,23 +366,20 @@ theorem mem_roots_map [CommRing k] [IsDomain k] {f : R →+* k} {x : k} (hp : p
 #align polynomial.mem_roots_map Polynomial.mem_roots_map
 
 -- Porting note: previously could not synthesize Algebra R S
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Polynomial/FieldDivision.lean:368]: Can these be removed?

File: Mathlib/Data/Polynomial/Module.lean [hunk 7]
```diff
@@ -201,7 +198,6 @@ theorem smul_apply (f : R[X]) (g : PolynomialModule R M) (n : ℕ) :
 -- possibly due to either lean4#2074 or upstream workarounds for it.
 -- One way to prune the bad branches is:
 attribute [-instance] IsDomain.toCancelCommMonoidWithZero in
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Polynomial/Module.lean:200]: Can this be removed?

File: Mathlib/Data/Polynomial/RingDivision.lean [hunk 8]
```diff
@@ -1206,7 +1205,6 @@ theorem card_roots_le_map_of_injective [IsDomain A] [IsDomain B] {p : A[X]} {f :
 This also works
 `attribute [-instance] Ring.toNonAssocRing`
 -/
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Polynomial/RingDivision.lean:1207]: Remove?

File: Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean [hunk 9]
```diff
@@ -93,17 +93,14 @@ theorem finrank_eq_card_chooseBasisIndex :
...
 @[simp]
 theorem finrank_finsupp {ι : Type v} [Fintype ι] : finrank R (ι →₀ R) = card ι := by
   rw [finrank, rank_finsupp_self, ← mk_toNat_eq_card, toNat_lift]
 #align finite_dimensional.finrank_finsupp FiniteDimensional.finrank_finsupp
 
-set_option synthInstance.etaExperiment true in
 /-- The finrank of `(ι → R)` is `Fintype.card ι`. -/
 theorem finrank_pi {ι : Type v} [Fintype ι] : finrank R (ι → R) = card ι := by
-  set_option synthInstance.etaExperiment true in -- Porting note: gets around lean4#2074
     simp [finrank]
```
Comment (Ruben-VandeVelde) [inline Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean:104]:
```suggestion
  simp [finrank]
```

File: Mathlib/LinearAlgebra/Lagrange.lean [hunk 10]
```diff
@@ -313,26 +307,22 @@ def interpolate (s : Finset ι) (v : ι → F) : @LinearMap _ _ _ _ (RingHom.id
 -- Porting note: The arguments for 'LinearMap' on line 301 had to be given explicitly because
 -- the instance 'Module F F[X]' could not be synthetized
```
Comment (Ruben-VandeVelde) [inline Mathlib/LinearAlgebra/Lagrange.lean:308]: Is this fixed now too?

File: Mathlib/Order/GameAdd.lean [hunk 11]
```diff
@@ -226,15 +226,18 @@ namespace Sym2
...
     (fun ⟨x₁, x₂⟩ IH' => IH x₁ x₂ fun a' b' => IH' ⟨a', b'⟩) (a, b)
 #align sym2.game_add.fix Sym2.GameAdd.fix
 
 theorem GameAdd.fix_eq {C : α → α → Sort _} (hr : WellFounded rα)
     (IH : ∀ a₁ b₁, (∀ a₂ b₂, Sym2.GameAdd rα ⟦(a₂, b₂)⟧ ⟦(a₁, b₁)⟧ → C a₂ b₂) → C a₁ b₁) (a b : α) :
-    GameAdd.fix hr IH a b = IH a b fun a' b' _ => GameAdd.fix hr IH a' b' :=
-  WellFounded.fix_eq _ _ _
+    GameAdd.fix hr IH a b = IH a b fun a' b' _ => GameAdd.fix hr IH a' b' := by
+  dsimp [GameAdd.fix]
+  exact WellFounded.fix_eq _ _ _
```
Comment (Ruben-VandeVelde) [inline Mathlib/Order/GameAdd.lean:240]: Do these need porting notes?

File: Mathlib/RingTheory/EisensteinCriterion.lean [hunk 12]
```diff
@@ -36,7 +36,8 @@ theorem map_eq_C_mul_X_pow_of_forall_coeff_mem {f : R[X]} {P : Ideal R}
     (hfP : ∀ n : ℕ, ↑n < f.degree → f.coeff n ∈ P) :
     map (mk P) f = C ((mk P) f.leadingCoeff) * X ^ f.natDegree :=
   Polynomial.ext fun n => by
-    by_cases hf0 : f = 0; · simp [hf0]
+    by_cases hf0 : f = 0;
```
Comment (Ruben-VandeVelde) [inline Mathlib/RingTheory/EisensteinCriterion.lean:39]:
```suggestion
    by_cases hf0 : f = 0
```

File: Mathlib/RingTheory/Ideal/Quotient.lean [hunk 13]
```diff
@@ -371,21 +367,19 @@ instance modulePi : Module (R ⧸ I) ((ι → R) ⧸ I.pi ι) where
...
 #align ideal.module_pi Ideal.modulePi
 
-set_option synthInstance.etaExperiment false in -- Porting note: needed, otherwise type times out
 /-- `R^n/I^n` is isomorphic to `(R/I)^n` as an `R/I`-module. -/
-noncomputable def piQuotEquiv : ((ι → R) ⧸ I.pi ι) ≃ₗ[R ⧸ I] ι → (R ⧸ I) := by
-  refine' ⟨⟨⟨?toFun, _⟩, _⟩, ?invFun, _, _⟩
-  case toFun => set_option synthInstance.etaExperiment true in -- Porting note: to get `Module R R`
-    exact fun x ↦
+noncomputable def piQuotEquiv : ((ι → R) ⧸ I.pi ι) ≃ₗ[R ⧸ I] ι → (R ⧸ I) where
+  toFun := fun x ↦
```
Comment (Ruben-VandeVelde) [inline Mathlib/RingTheory/Ideal/Quotient.lean:372]: If you're reformatting all this, might as well use
```suggestion
  toFun x :=
```
(and similar for other fields)

File: Mathlib/Topology/Algebra/Module/Basic.lean [hunk 14]
```diff
@@ -2484,7 +2410,6 @@ def unitsEquiv : (M →L[R] M)ˣ ≃* M ≃L[R] M where
 #align continuous_linear_equiv.units_equiv ContinuousLinearEquiv.unitsEquiv
 
 -- Porting note: mismatched types
```
Comment (Ruben-VandeVelde) [inline Mathlib/Topology/Algebra/Module/Basic.lean:2412]: Is this still relevant?

File: Mathlib/RingTheory/Ideal/Quotient.lean (hunk 13 above)
Comment (semorrison) [reply Mathlib/RingTheory/Ideal/Quotient.lean:372]: Not today. :-)

File: Mathlib/LinearAlgebra/Lagrange.lean (hunk 10 above)
Comment (semorrison) [reply Mathlib/LinearAlgebra/Lagrange.lean:308]: Fixed!

Comment (semorrison) [reply Mathlib/LinearAlgebra/Lagrange.lean:308]: Fixed!

File: Mathlib/RingTheory/Ideal/Quotient.lean (hunk 13 above)
Comment (semorrison) [reply Mathlib/RingTheory/Ideal/Quotient.lean:372]: Not today. :-)

**Comment by semorrison** [comment]
I have a clean build/test/lint locally, so:

bors p=10

bors merge

**Pull request closed without merging**